Fixes #21170 - move password strength meter logic to react

amirfefer opened the pull request:
`password_strength.js` uses jquery_pwstrength_bootstrap wrapper 
I've added `react-password-strength` - react component,  which renders password input with integrated password strength meter by using [zxcvbn](https://github.com/dropbox/zxcvbn) to calculate a password strength score.
In addition, I moved `verify password` logic to wepback (it was  in `password_strength` file as well)
Tests are missing for now.

_after update:_
Try online with [Stroybook](http://rawgit.com/amirfefer/foreman/storybook-my-feature/index.html?selectedKind=Password%20Strength%20Meter&selectedStory=With%20no%20error%20&full=0&down=1&left=1&panelRight=0&downPanel=storybook%2Factions%2Factions-panel) 

_before update:_
![oct 3 2017 4-14 pm - edited 1](https://user-images.githubusercontent.com/11807069/31127096-8c166150-a856-11e7-8535-0ffdd109c9f6.gif)

amirfefer committed:
Fixes #21170 - move password_strength.js to wepback

amirfefer commented:
@rohoover What do you say about the color scheme?

Rohoover commented:
I think it jumps from orange to green. Maybe a yellow in between for 

Weak - Red
Normal - Orange
Medium -Yellow
Strong -Green
Very Strong - Dark Green

waldenraines commented on app/views/users/_form.html.erb:
```diff
@@ -41,8 +40,13 @@
         <% end %>
-        <%= password_f f, :password, :label => _('Password'), :error => @user.errors[:password_hash] %>
-        <%= password_f f, :password_confirmation, :label => _('Verify') %>
+        <div class="clearfix" id="password_react"></div>
```
Is the `.clearfix` needed here?

waldenraines commented on webpack/assets/javascripts/react_app/components/passwordStrength/index.js:
```diff
@@ -0,0 +1,31 @@
+
+    return (
+      <div className={`form-group ${error.length > 0 ? 'has-error' : ''}`}>
+        <label className="col-md-2 control-label"> { __('Password') } </label>
```
This label should include a `for` attribute to tie it to the input below for accessibility.

ohadlevy replied:
I suggest to use the CommonForm component instead of redefining it here (https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/forms/CommonForm.js)

waldenraines commented on webpack/assets/javascripts/react_app/components/passwordStrength/password_strength.scss:
```diff
@@ -0,0 +1,10 @@
+.ReactPasswordStrength-input {
+  width: 100% !important;
```
Can we style this without the `!important`s?  Maybe add additional classes or elements to the selector if you need more specificity.

ohadlevy commented on webpack/assets/javascripts/react_app/components/passwordStrength/index.js:
```diff
@@ -0,0 +1,31 @@
+        </div>
+        { error.length > 0 &&
+        <span className="help-block help-inline">
+          <span className="error-message"> {error[0]} </span>
```
could there be more than one error? should we only display the first one?

amirfefer commented:
@waldenraines  I've mentioned your comments.
@ohadlevy, I've switched verify password validation logic from jquery to react

I've changed color pattern to @Rohoover  suggestion

waldenraines reviewed (approved):
Looks good to me!

amirfefer commented:
@Rohoover 
thoughts about color scheme , text placement ?
![oct 19 2017 11-28 pm - edited](https://user-images.githubusercontent.com/11807069/31792763-cfde2cd4-b525-11e7-8d2f-43a2e4c785d1.gif)

ares commented:
Copying out notes from UX demo with @Rohoover about this:

> Password authorization - Weak/Normal/Strong visualization would not pass 528 compliance for those visually impaired. The colors associated with each strength level seems mismatched. For example: Normal is red. A red status as well as form field highlight indicates an error. Weak is grey, and also possibly not high enough contrast for the visually impaired. The solution may be to simplify the password strength to text only, “Weak Password, Strong Password”. Also - greater thought or discussion could be had around “Normal” or “Strong”. If a “Normal” indicator is shown, does this actually change user behavior in the same way “Weak” does? Perhaps only “weak” is needed.

While I think most of it is addressed and it's a great improvement already, I think there are some other things that could be considered while it's being touched.
* If messages include icons, it would solve the coloring issue.
* fonts seems to be different for pass match and pass strength (italic, different size)
* "too short" should be perhaps "Too short" to be consistent
* does medium password still get accepted?

Also when this gets in, please update/close http://projects.theforeman.org/issues/20740 which tracked this originally :-)

amirfefer commented:
Thanks @ares 
I've re-styled it a bit:
![oct 20 2017 2-39 pm - edited](https://user-images.githubusercontent.com/11807069/31819424-34bb4978-b5a5-11e7-8da8-5fd5af52e36e.gif)

amirfefer commented on webpack/assets/javascripts/react_app/components/passwordStrength/index.js:
```diff
@@ -0,0 +1,51 @@
+            changeCallback={(data) => {this.setState({password: data.password});}}
+            minLength={6}
+            minScore={2}
+            userInputs={[username]}
```
@ohadlevy, `user_inputs` array adds strings to [zxcvbn](https://github.com/dropbox/zxcvbn) dictionary.  its meant for user inputs from other fields of the form, so this PR  doesn't break username reference.

ohadlevy replied:
this only works if the username is entered prior to the initializing the component, you would need to bind to the form via js not via a react mounter props.

amirfefer replied:
@ohadlevy  , the username value isn't a prop from react mounter, I'm using the DOM  dynamically when rendering, as you can see in line 25.

ohadlevy replied:
that only reads it when the component renders, it should be an event of when it changes....

ohadlevy reviewed (changes requested):
Please:
1. switch to redux instead of local state
2. bind to the user input in the form instead of receiving it as a prop.
3. add tests :-)
4. would be to vet it with @Rohoover hoover again 

thanks!

Rohoover commented:
I think the color bar is a big improvement. Also - the text is good because it provides something other than just a color state. I'm a little concerned the the yellow color may not provide enough contrast, but this solution is great overall so I'm not worried if it can't be addressed immediately.

ohadlevy commented:
@amirfefer can you please rebase?

amirfefer commented:
@ohadlevy, I've rebased, could you have another look?

ohadlevy commented:
@sharvit can you please review?

sharvit commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/index.js:
```diff
@@ -0,0 +1,99 @@
+
+const mapDispatchToProps = dispatch => bindActionCreators(
+  {
+    changePassword: changePass,
```
Can't it just be `changePassword` everywhere?
I would prefer tp name it as `updatePassword`.

sharvit commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/index.js:
```diff
@@ -0,0 +1,99 @@
+          <ReactPasswordStrength
+            changeCallback={(data) => {
+              changePassword(data.password);
+            }}
```
Can be shorter
```js
changeCallback={({ password }) => changePassword(password)}
```

sharvit commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/index.js:
```diff
@@ -0,0 +1,99 @@
+class PasswordStrength extends React.Component {
+  constructor(props) {
+    super(props);
+    this.checkRetype = this.checkRetype.bind(this);
```
Should use the helper here
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/common/helpers.js#L2-L7

sharvit commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/index.js:
```diff
@@ -0,0 +1,99 @@
+    this.state = {
+      password: '',
+      error: props.data.verify.error,
+    };
```
I don't see where you are using this state,
maybe it should be remove because you mapping redux state to props?

sharvit commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/index.js:
```diff
@@ -0,0 +1,99 @@
+              userInputIds.length > 0
+                ? userInputIds.map(input => document.getElementById(input).value)
+                : []
+            }
```
I am not sure i understand the meaning of that, can you explain please?

Also, haven't found this `#user_login` all over the project, what is it exactly?

```js
:userInputIds => ['user_login'],
```

amirfefer replied:
Array of input ids to add to `zxcvbn's` dictionary.
This way, the password score is decreased if it composite from other inputs such as username.

sharvit commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/index.js:
```diff
@@ -0,0 +1,99 @@
+    } else {
+      this.props.passwordMatch(true);
+    }
+  }
```
I don't think it should be a part of the component to decide if the passwords are match.

What do you think about:
```js
// actions
updatePassword, updatePasswordConfirmation

// state
{ password: '', passwordConfirmation: '', match: true }

// reducer
// { payload: { password: 'pass' } } or
// { payload: { passwordConfirmation: 'pass '} }
case PASSWORD_STRENGTH_PASSWROD_UPDATED:
case PASSWORD_STRENGTH_PASSWROD_CONFIRMATION_UPDATED:
  state = state.merge(payload);
  return state.set('match', state.password === state.passwordConfirmation);
```

sharvit commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/password_strength.scss:
```diff
@@ -0,0 +1,75 @@
+
+  .is-strength-0 {
+    .ReactPasswordStrength-strength-desc {
+      color: #a30000;
```
Can you use variables for those colors?

sharvit commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/password_strength.scss:
```diff
@@ -0,0 +1,75 @@
+  .ReactPasswordStrength-input {
+    width: 100%;
+    font-size: 12px;
+    font-family: 'Open Sans', Helvetica, Arial, sans-serif;
```
We really need to change to font here? Isn't it the default font?

amirfefer replied:
The component default font is different from foreman default

sharvit commented on webpack/assets/javascripts/react_app/redux/consts.js:
```diff
@@ -24,3 +24,5 @@ export const NOTIFICATIONS_MARK_GROUP_AS_READ =
 export const NOTIFICATIONS_POLLING_STARTED =
   'NOTIFICATIONS_POLLING_STARTED';
+export const PASSWROD_CHANGED = 'PASSWROD_CHANGED';
+export const PASSWORD_MATCHED = 'PASSWORD_MATCHED';
```
I think it is better to use here a more contextual names like `PASSWORD_STRENGTH_PASSWROD_CHANGED`

sharvit commented on webpack/stories/index.js:
```diff
@@ -1,3 +1,4 @@
+
```
Just remove this line, there is no updates for this file...

sharvit commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/index.js:
```diff
@@ -0,0 +1,99 @@
+              __('Very strong'),
+            ]}
+            inputProps={{
+              autoComplete: 'new-password',
```
What is this `autoComplete: 'new-password'`?
Can you please explain?

amirfefer commented:
@sharvit Thanks for your review. could you have a second look?

sharvit commented:
@amirfefer nice work!

1. I think that your actions should be `updatePassword` and `updatePasswordConfirmation` because this is what the user can actually do.
2. Since there is no inner state, your component can be a dummy/functional component.

amirfefer commented:
@sharvit  Thanks again !
about making it a functional component,  well, we'll lose react life cycle rendering and because we're using redux here, i think it better to keep it as a class component.
The other way it to split it to two components: a stateless functional component, and a container component which connect to redux. imo those changes are redundant.

amirfefer commented:
@ohadlevy, @waldenraines , @sharvit 
Could we merge this?

sharvit reviewed (approved):
Good to me! 👍

ohadlevy commented:
@amirfefer please rebase.

ohadlevy commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/passwordStrength.test.js:
```diff
@@ -0,0 +1,62 @@
+  passwordNotMatched,
+} from './passwordStrength.fixtures';
+
+configure({ adapter: new Adapter() });
```
these setup are no longer need (see webpack/test_setup.js)

ohadlevy commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/passwordStrength.test.js:
```diff
@@ -0,0 +1,62 @@
+configure({ adapter: new Adapter() });
+
+jest.unmock('./');
+global.__ = jest.fn();
```
not required anymore

ohadlevy commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/passwordStrength.test.js:
```diff
@@ -0,0 +1,62 @@
+function setup(verify, error, state) {
+  const store = mockStore(state);
+
+  return mount(<PasswordStrength
```
would shallow work here too?

amirfefer replied:
shallow doesn't render inner components therefore the resulted snapshot lacks in inner content

ohadlevy commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/index.js:
```diff
@@ -0,0 +1,88 @@
+            inputProps={{
+              name,
+              id,
+              className: `${className}`,
```
do you really need the string interrelation here? I think the following would just work:
```js
inputProps={{  name,  id, className }}
```

ohadlevy commented on webpack/assets/javascripts/react_app/components/user/passwordStrength/index.js:
```diff
@@ -0,0 +1,88 @@
+          <CommonForm
+            label={__('Verify')}
+            touched={true}
+            error={matchMessage ? verify.error : __('Password do not matched')}
```
The original string was ```passwords do not match```

ohadlevy reviewed (changes requested):
Initial review looks good - I've added a few comments inline.

Please also adds action/reducers tests.

Thanks!

amirfefer commented:
@ohadlevy  thanks, I've fixed and rebased.  would you mind take another look?

mmoll commented:
@amirfefer Is the test failure in the foreman test related to these changes?

ares commented:
ping @amirfefer

amirfefer commented:
I've fixed the users integration test, 
Apparently the react component did not render at all via capybara.

mmoll commented:
@amirfefer please rebase

I guess the travis test "failures" also need fixing/updating...

amirfefer commented:
Rebased.
plus I've updated jest snapshot.

ohadlevy commented:
@amirfefer anything left here?

ohadlevy commented on webpack/assets/javascripts/react_app/common/colors.scss:
```diff
@@ -1 +1,7 @@
 $pf-black-200: #ededed;
+$pf-red-200: #a30000;
```
can we use this from the patternfly package? @sharvit ?

sharvit replied:
When my css pr will merge you would be able to import variables:
https://github.com/theforeman/foreman/pull/5212/files#diff-f430da9fad5e785cb1f815fd719fdc5aR1

ohadlevy commented:
@amirfefer can you please also include
```difff
diff --git a/config/webpack.vendor.js b/config/webpack.vendor.js
index aed3a4bde..e6e2f023a 100644
--- a/config/webpack.vendor.js
+++ b/config/webpack.vendor.js
@@ -9,6 +9,7 @@ module.exports = [
   'react-ellipsis-with-tooltip',
   'react-numeric-input',
   'react-onclickoutside',
+  'react-password-strength',
   'patternfly-react',
   'react-redux',
   'redux',
```

amirfefer commented:
@ohadlevy - done :)

amirfefer commented:
[test foreman]

sharvit reviewed (approved):
LGTM!

amirfefer commented:
@ohadlevy tests are green :)
